### PR TITLE
Feat: create relations between trades and transactions

### PR DESCRIPTION
**Pull request by josepfo**

### CODE REVIEW

**Comment by josepfo** [comment]
[Record internal TX UUIDs in trades table, relay via API to exchanges' user reference fields](https://trello.com/c/orPhOEDP/642-record-internal-tx-uuids-in-trades-table-relay-via-api-to-exchanges-user-reference-fields)

File: lib/plugins.js [hunk 1]
```diff
@@ -533,8 +552,9 @@ function plugins (settings, deviceId) {
       _.mapKeys(_.snakeCase)
     )
     const tradeEntry = massage(_tradeEntry, error)
-    const sql = pgp.helpers.insert(tradeEntry, null, 'trades')
-    return db.none(sql)
+    const sql = pgp.helpers.insert(tradeEntry, null, 'trades') + 'RETURNING *'
+    return db.oneOrNone(sql)
+      .then(newTrade => recordTradeAndTx(newTrade.id, _tradeEntry.internalTxIdList))
```
Comment (hellypi) [inline lib/plugins.js]: This also looks like chained queries, if they are, do them inside a task or a tx (transaction) https://vitaly-t.github.io/pg-promise/Task.html

**Review by hellypi — CHANGES_REQUESTED** [review]

**Review by hellypi — APPROVED** [review]

File: lib/plugins.js [hunk 2]
```diff
@@ -428,12 +432,15 @@ function plugins (settings, deviceId) {
 
     if (filtered.length === 0) return null
 
+    const internalTxIdList = _.uniqBy('txId', _.map(tradeElement => { return { txId: tradeElement.internalTxId, direction: tradeElement.direction } })(filtered))
+
```
Comment (RafaelTaranto) [inline lib/plugins.js]:
```suggestion
    const partitionByDirection = _.partition(({ direction }) => direction === 'cashIn'))
    const [cashInTxs, cashOutTxs] = _.compose(partitionByDirection, _.uniqBy('internalTxId'))(filtered)
```

File: lib/plugins.js [hunk 3]
```diff
@@ -428,12 +432,15 @@ function plugins (settings, deviceId) {
...
     const cryptoAtoms = filtered
       .reduce((prev, current) => prev.plus(current.cryptoAtoms), BN(0))
 
     const timestamp = filtered.map(r => r.timestamp).reduce((acc, r) => Math.max(acc, r), 0)
 
     const consolidatedTrade = {
+      internalTxIdList,
```
Comment (RafaelTaranto) [inline lib/plugins.js]:
```suggestion
      cashInTxs,
      cashOutTxs,
```

File: lib/plugins.js [hunk 4]
```diff
@@ -533,8 +567,13 @@ function plugins (settings, deviceId) {
       _.mapKeys(_.snakeCase)
     )
     const tradeEntry = massage(_tradeEntry, error)
-    const sql = pgp.helpers.insert(tradeEntry, null, 'trades')
-    return db.none(sql)
+    const sql = pgp.helpers.insert(tradeEntry, null, 'trades') + 'RETURNING *'
+    db.tx(async t => {
+      const newTrade = await t.oneOrNone(sql)
+      return recordTradeAndTx(newTrade.id, _tradeEntry.internalTxIdList, t)
```
Comment (RafaelTaranto) [inline lib/plugins.js]:
```suggestion
      return recordTradeAndTx(newTrade.id, _tradeEntry, t)
```

File: lib/plugins.js [hunk 5]
```diff
@@ -533,8 +567,13 @@ function plugins (settings, deviceId) {
...
+    })
+      .then(data => data)
+      .catch(error => error)
```
Comment (RafaelTaranto) [inline lib/plugins.js]: This looks useless.

File: lib/plugins.js [hunk 6]
```diff
@@ -533,8 +567,13 @@ function plugins (settings, deviceId) {
       _.mapKeys(_.snakeCase)
     )
     const tradeEntry = massage(_tradeEntry, error)
-    const sql = pgp.helpers.insert(tradeEntry, null, 'trades')
-    return db.none(sql)
+    const sql = pgp.helpers.insert(tradeEntry, null, 'trades') + 'RETURNING *'
+    db.tx(async t => {
+      const newTrade = await t.oneOrNone(sql)
```
Comment (RafaelTaranto) [inline lib/plugins.js]: don't use async on the server for now

File: lib/plugins.js [hunk 7]
```diff
@@ -504,6 +511,33 @@ function plugins (settings, deviceId) {
       })
   }
 
+  function recordTradeAndTx (tradeId, internalTxIdList, dbTx) {
```
Comment (RafaelTaranto) [inline lib/plugins.js]:
```suggestion
  function recordTradeAndTx (tradeId, { cashInTxs, cashOutTxs }, dbTx) {
```

File: migrations/1616687417371-add_internal_tx_to_trades.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+exports.up = function (next) {
+  var sql = [
+    `CREATE TABLE cashout_tx_trades (
+      cash_out_tx_id uuid REFERENCES cash_out_txs(id),
+      trade_id serial REFERENCES trades(id),
+      CONSTRAINT cashout_trade_pkey PRIMARY KEY (cash_out_tx_id,trade_id)
+    )`,
+    `CREATE TABLE cashin_tx_trades (
+      cash_in_tx_id uuid REFERENCES cash_in_txs(id),
```
Comment (RafaelTaranto) [inline migrations/1616687417371-add_internal_tx_to_trades.js]:
```suggestion
      tx_id uuid REFERENCES cash_in_txs(id),
```

File: migrations/1616687417371-add_internal_tx_to_trades.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+const db = require('./db')
+
+exports.up = function (next) {
+  var sql = [
+    `CREATE TABLE cashout_tx_trades (
+      cash_out_tx_id uuid REFERENCES cash_out_txs(id),
```
Comment (RafaelTaranto) [inline migrations/1616687417371-add_internal_tx_to_trades.js]:
```suggestion
      tx_id uuid REFERENCES cash_out_txs(id),
```

File: lib/plugins.js [hunk 10]
```diff
@@ -504,6 +511,33 @@ function plugins (settings, deviceId) {
...
+      }
+    }, internalTxIdList)
+    if (!_.isEmpty(cashinTxs)) {
+      const query = pgp.helpers.insert(cashinTxs, columnSetCashin)
+      queries.push(dbTx.none(query))
+    }
+    if (!_.isEmpty(cashoutTxs)) {
+      const query = pgp.helpers.insert(cashoutTxs, columnSetCashout)
+      queries.push(dbTx.none(query))
+    }
```
Comment (RafaelTaranto) [inline lib/plugins.js]:
```js
  const columnSetCashIn = new pgp.helpers.ColumnSet(['tx_id', 'trade_id'], { table: 'cashin_tx_trades' })
  const columnSetCashOut = new pgp.helpers.ColumnSet(['tx_id', 'trade_id'], { table: 'cashout_tx_trades' })

  const mapToEntry = _.map((tx, tradeId) => ({ tx_id: tx.internalTxId, trade_id: tradeId }))
  const queries = []

  if (!_isEmpty(cashInTxs)) {
    const query = pgp.helpers.insert(mapToEntry(cashInTxs), columnSetCashIn)
    queries.push(dbTx.none(query))
  }

  if (!_isEmpty(cashOutTxs)) {
    const query = pgp.helpers.insert(mapToEntry(cashOutTxs), columnSetCashOut)
    queries.push(dbTx.none(query))
  }
```

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

**Review by RafaelTaranto — APPROVED** [review]

**Pull request merged**